addons/binding/org.openhab.binding.amazonechocontrol/src/main/java/org/openhab/binding/amazonechocontrol/handler/AccountHandler.java

```diff
@@ -0,0 +1,554 @@
+        }
+    }
+
+    public @Nullable Device findDeviceJson(EchoHandler echoHandler) {
+        String serialNumber = echoHandler.findSerialNumber();
+        return findDeviceJson(serialNumber);
+    }
+
+    public @Nullable Device findDeviceJson(@Nullable String serialNumber) {
+        Device result = null;
+        if (StringUtils.isNotEmpty(serialNumber)) {
+            Map<String, Device> jsonSerialNumberDeviceMapping = this.jsonSerialNumberDeviceMapping;
+            result = jsonSerialNumberDeviceMapping.get(serialNumber);
+        }
+        return result;
+    }
+
+    public @Nullable Device findDeviceJsonBySerialOrName(@Nullable String serialOrName) {
+        if (StringUtils.isNotEmpty(serialOrName)) {
+            Map<String, Device> currentJsonSerialNumberDeviceMapping = this.jsonSerialNumberDeviceMapping;
+            for (Device device : currentJsonSerialNumberDeviceMapping.values()) {
+                if (StringUtils.equalsIgnoreCase(device.serialNumber, serialOrName)) {
+                    return device;
+                }
+            }
+            for (Device device : currentJsonSerialNumberDeviceMapping.values()) {
+                if (StringUtils.equalsIgnoreCase(device.accountName, serialOrName)) {
+                    return device;
+                }
+            }
+        }
+        return null;
+    }
+
+    @Override
+    public void updateDeviceList(boolean manualScan) {
+        if (manualScan) {
+            discoverFlashProfiles = true;
+        }
+
+        Connection currentConnection = connection;
+        if (currentConnection == null) {
+            return;
+        }
+
+        List<Device> devices = null;
+        try {
+            if (currentConnection.getIsLoggedIn()) {
+                devices = currentConnection.getDeviceList();
+            }
+        } catch (IOException | URISyntaxException e) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, e.getLocalizedMessage());
+        }
+        if (devices != null) {
+            Map<String, Device> newJsonSerialDeviceMapping = new HashMap<>();
+            for (Device device : devices) {
+                String serialNumber = device.serialNumber;
+                if (serialNumber != null) {
+                    newJsonSerialDeviceMapping.put(serialNumber, device);
+                }
+            }
+            jsonSerialNumberDeviceMapping = newJsonSerialDeviceMapping;
+            amazonEchoDiscovery.setDevices(getThing().getUID(), devices);
+        }
+        synchronized (echoHandlers) {
+            for (EchoHandler child : echoHandlers) {
+                initializeEchoHandler(child, currentConnection);
+            }
+        }
+
+        updateFlashBriefingHandlers(currentConnection);
+    }
+
+    public void setEnabledFlashBriefingsJson(String flashBriefingJson) {
+        Connection currentConnection = connection;
+        Gson gson = new Gson();
```

Comment:
This is potentially expensive - you should instantiate it only once as a field in that class.